frontend/app/services/raoidc-service.server.ts

```diff
@@ -0,0 +1,113 @@
+      return [];
+    }
+
+    const jwk = await subtle.exportKey('jwk', await publicKeyPemToCryptoKey(AUTH_JWT_PUBLIC_KEY));
+    return [{ ...jwk, kid: generateJwkId(jwk) } as JsonWebKey & { kid: string }];
+  }
+
+  /**
+   * Generates an OAuth authentication request. Used to kickstart the OAuth login process.
+   */
+  function generateAuthRequest(redirectUri: string) {
+    const { codeChallenge, codeVerifier } = generateCodeChallenge();
+    const clientId = AUTH_RAOIDC_CLIENT_ID;
+    const scope = 'openid profile';
+    const state = generateRandomState();
+
+    const authUrl = generateAuthorizationRequest(serverMetadata.authorization_endpoint, clientId, codeChallenge, redirectUri, scope, state);
+
+    return { authUrl, codeVerifier, state };
+  }
+
+  /**
+   * Handle an OAuth login callback.
+   */
+  async function handleCallback(request: Request, codeVerifier: string, expectedState: string, redirectUri: string) {
+    const authCode = new URL(request.url).searchParams.get('code');
+    const error = new URL(request.url).searchParams.get('error');
+    const state = new URL(request.url).searchParams.get('state');
+
+    if (error) {
+      throw new Error(`Unexpected error: ${error}`);
+    }
+
+    if (!authCode) {
+      throw new Error('Missing authcode in response');
+    }
+
+    if (state !== expectedState) {
+      throw new Error('CSRF error: state does not match');
+    }
+
+    const privateCryptoKey = await privateKeyPemToCryptoKey(AUTH_JWT_PRIVATE_KEY!);
+    const privateKeyId = generateJwkId(await subtle.exportKey('jwk', privateCryptoKey));
+
+    const client: ClientMetadata = {
+      clientId: AUTH_RAOIDC_CLIENT_ID,
+      privateKey: privateCryptoKey,
+      privateKeyId: privateKeyId,
+    };
+
+    const authTokenSet = await fetchAccessToken(serverMetadata, serverJwkSet, authCode, client, codeVerifier, redirectUri, fetchFn);
+    const userInfo = await fetchUserInfo(serverMetadata.userinfo_endpoint, authTokenSet.access_token, client, fetchFn);
+
+    return { auth: authTokenSet, user_info: userInfo };
+  }
+
+  /**
+   * Generate a key id from the modulus of a jwk.
+   */
+  function generateJwkId(jwk: JsonWebKey) {
+    return createHash('md5')
+      .update(jwk.n ?? '')
+      .digest('hex');
+  }
+
+  function getFetchFn(proxyUrl?: string) {
+    if (proxyUrl) {
+      return async (input: string | URL, init?: FetchFunctionInit) => {
+        const dispatcher = new ProxyAgent({ uri: proxyUrl, proxyTls: { timeout: 30000 } });
+        const response = await undiciFetch(input, { ...init, dispatcher });
+        return new Response(toNodeReadable(response.body));
+      };
+    }
+
+    return async (input: string | URL, init?: FetchFunctionInit) => fetch(input, init);
+  }
+
+  return { getPublicJwks, generateAuthRequest, handleCallback };
+}
+
+// TODO :: GjB :: this seems wrong; is there a better way to avoid creating the service if AUTH_ENABLED is false?
```

Comment:
Is the `createRaoidcService()` should throw an `Error` if `env` variables are not configured correctly. Otherwise you need to do this everywhere.

```tsx
if (!raoidcService) {
    log.warn('Call to /auth/callback/%s but authentication is disabled', params.providerId);
    // TODO :: GjB :: handle this better than just throwing an error
    throw new Error('RAOIDC service not configured');
  }
```

Reply:
Let me try to implement your suggestion to see how the application behaves.

Reply:
I'm running into issues when I throw inside of the `createRaoidcService()` function because of how ESBuild hoists all of the ESM imports up to the global scope.

For now, are you okay to merge this as-is and create a task to clean up how the `AUTH_ENABLED` feature flag is handled in a later PR?